q2_quality_filter/_filter.py

```diff
@@ -6,181 +6,448 @@
-                writer = gzip.open(str(path), mode='w')
-            writer.write(fastq_lines)
-
-            log_records_totalkept_counts[sample_id] += 1
-
-        if writer is not None:
-            manifest_fh.write('%s,%s,%s\n' % (sample_id, path.name, 'forward'))
-            writer.close()
-
-    if set(log_records_totalkept_counts.values()) == {0, }:
-        raise ValueError("All sequences from all samples were filtered out. "
-                         "The parameter choices may be too stringent for the "
-                         "data.")
-
+    phred_offset = yaml.load(
+        metadata_view, Loader=yaml.SafeLoader
+    )['phred-offset']
+    demux_manifest_df = demux.manifest.view(pd.DataFrame)
+
+    # initialize filtering stats tracking dataframe
+    filtering_stats_df = pd.DataFrame(
+        data=0,
+        index=demux_manifest_df.index,
+        columns=[
+            'total-input-reads',
+            'total-retained-reads',
+            'reads-truncated',
+            'reads-too-short-after-truncation',
+            'reads-exceeding-maximum-ambiguous-bases'
+        ]
+    )
+
+    for barcode_id, sample_id in enumerate(demux_manifest_df.index.values):
+        # get/create filepath(s) of input/output fastq file(s)
+        forward_input_fp = demux_manifest_df.loc[sample_id, 'forward']
+        forward_output_fp = Path(result.path) / Path(forward_input_fp).name
+        if paired:
+            reverse_input_fp = demux_manifest_df.loc[sample_id, 'reverse']
+            reverse_output_fp = Path(result.path) / Path(reverse_input_fp).name
+
+        # open output filehandle(s) and create fastq record iterator
+        forward_fh = gzip.open(forward_output_fp, mode='wb')
+
+        if paired:
+            reverse_fh = gzip.open(reverse_output_fp, mode='wb')
+
+            forward_iterator = _read_fastq_records(str(forward_input_fp))
+            reverse_iterator = _read_fastq_records(str(reverse_input_fp))
+            iterator = zip(forward_iterator, reverse_iterator)
+        else:
+            iterator = _read_fastq_records(str(forward_input_fp))
+
+        for fastq_record in iterator:
+            if paired:
+                forward_record, reverse_record = fastq_record
+            else:
+                forward_record = fastq_record
+                reverse_record = None
+
+            # process records
+            forward_record, forward_status = _process_record(
+                fastq_record=forward_record,
+                phred_offset=phred_offset,
+                min_quality=min_quality,
+                window_length=quality_window + 1,
+                min_length_fraction=min_length_fraction,
+                max_ambiguous=max_ambiguous
+            )
+            reverse_record, reverse_status = _process_record(
+                fastq_record=reverse_record,
+                phred_offset=phred_offset,
+                min_quality=min_quality,
+                window_length=quality_window + 1,
```

Comment:
Why the `+1` here?

Reply:
Due to the cli specified interpretation of `quality_window` compared to what the function searches for. `quality_window` is defined as the longest non-filtered window and `window_length` is defined as the shortest filtered window. I think the former interpretation is non-intuitive but didn't change so as not to make an interface change